File: apps/websockets/src/namespaces/obs.ts
import { Empty } from '@tsuwari/grpc/generated/websockets/google/protobuf/empty';
import {
  ObsAudioDecreaseVolumeMessage, ObsAudioDisableOrEnableMessage,
  ObsAudioIncreaseVolumeMessage,
  ObsAudioSetVolumeMessage,
  ObsSetSceneMessage, ObsStopOrStartStream,
  ObsToggleAudioMessage,
  ObsToggleSourceMessage,
} from '@tsuwari/grpc/generated/websockets/websockets';
import SocketIo from 'socket.io';

import { authMiddleware, io } from '../libs/io.js';

const sockets: Map<string, SocketIo.Socket> = new Map();
export const obsNameSpace = io.of('obs');

obsNameSpace.use(authMiddleware);
obsNameSpace.on('connection', async (socket) => {
  const channelId = socket.handshake.auth.channelId;
  sockets.set(channelId, socket);
});

export const onSetAudio = async (data: ObsAudioSetVolumeMessage): Promise<Empty> => {
  const socket = sockets.get(data.channelId);
  if (!socket) return {};

  socket.emit('setVolume', data);

  return {};
};

export const onAudioIncrease = async (data: ObsAudioIncreaseVolumeMessage): Promise<Empty> => {
  const socket = sockets.get(data.channelId);
  if (!socket) return {};

  socket.emit('increaseVolume', data);

  return {};
};

export const onAudioDecrease = async (data: ObsAudioDecreaseVolumeMessage): Promise<Empty> => {
  const socket = sockets.get(data.channelId);
  if (!socket) return {};

  socket.emit('decreaseVolume', data);

  return {};
};

export const onToggleAudioSource = async (data: ObsToggleAudioMessage): Promise<Empty> => {
  const socket = sockets.get(data.channelId);
  if (!socket) return {};

  socket.emit('toggleAudioSource', data);

  return {};
};

export const onSetScene = async (data: ObsSetSceneMessage): Promise<Empty> => {
  const socket = sockets.get(data.channelId);
  if (!socket) return {};

  socket.emit('setScene', data);

  return {};
};

export const onToggleSource = async (data: ObsToggleSourceMessage): Promise<Empty> => {
  const socket = sockets.get(data.channelId);
  if (!socket) return {};

  socket.emit('toggleSource', data);

  return {};
};

export const onAudioEnable = async (data: ObsAudioDisableOrEnableMessage): Promise<Empty> => {
  const socket = sockets.get(data.channelId);
  if (!socket) return {};

  socket.emit('enableAudio', data);

  return {};
};

export const onAudioDisable = async (data: ObsAudioDisableOrEnableMessage): Promise<Empty> => {
  const socket = sockets.get(data.channelId);
  if (!socket) return {};

  socket.emit('disableAudio', data);

  return {};
};

export const onStartStream = async (data: ObsStopOrStartStream): Promise<Empty> => {
  const socket = sockets.get(data.channelId);
  if (!socket) return {};

  socket.emit('startStream', data);

  return {};
};

export const onStopStream = async (data: ObsStopOrStartStream): Promise<Empty> => {
  const socket = sockets.get(data.channelId);
  if (!socket) return {};

  socket.emit('stopStream', data);

  return {};
};

File: apps/websockets/internal/grpc_impl/obs.go
package grpc_impl

import (
	"context"

	"github.com/satont/twir/libs/grpc/generated/websockets"
	"google.golang.org/protobuf/types/known/emptypb"
)

func (c *GrpcImpl) ObsSetScene(
	_ context.Context,
	msg *websockets.ObsSetSceneMessage,
) (*emptypb.Empty, error) {
	if err := c.obsServer.SendEvent(msg.ChannelId, "setScene", msg); err != nil {
		return nil, err
	}

	return &emptypb.Empty{}, nil
}

func (c *GrpcImpl) ObsToggleSource(
	_ context.Context,
	msg *websockets.ObsToggleSourceMessage,
) (*emptypb.Empty, error) {
	if err := c.obsServer.SendEvent(msg.ChannelId, "toggleSource", msg); err != nil {
		return nil, err
	}

	return &emptypb.Empty{}, nil
}
func (c *GrpcImpl) ObsToggleAudio(
	_ context.Context,
	msg *websockets.ObsToggleAudioMessage,
) (*emptypb.Empty, error) {
	if err := c.obsServer.SendEvent(msg.ChannelId, "toggleAudioSource", msg); err != nil {
		return nil, err
	}

	return &emptypb.Empty{}, nil
}
func (c *GrpcImpl) ObsAudioSetVolume(_ context.Context, msg *websockets.ObsAudioSetVolumeMessage) (
	*emptypb.Empty, error,
) {
	if err := c.obsServer.SendEvent(msg.ChannelId, "setVolume", msg); err != nil {
		return nil, err
	}

	return &emptypb.Empty{}, nil
}
func (c *GrpcImpl) ObsAudioIncreaseVolume(
	_ context.Context, msg *websockets.ObsAudioIncreaseVolumeMessage,
) (*emptypb.Empty, error) {
	if err := c.obsServer.SendEvent(msg.ChannelId, "increaseVolume", msg); err != nil {
		return nil, err
	}

	return &emptypb.Empty{}, nil
}
func (c *GrpcImpl) ObsAudioDecreaseVolume(
	_ context.Context, msg *websockets.ObsAudioDecreaseVolumeMessage,
) (*emptypb.Empty, error) {
	if err := c.obsServer.SendEvent(msg.ChannelId, "decreaseVolume", msg); err != nil {
		return nil, err
	}

	return &emptypb.Empty{}, nil
}
func (c *GrpcImpl) ObsAudioEnable(
	_ context.Context,
	msg *websockets.ObsAudioDisableOrEnableMessage,
) (
	*emptypb.Empty, error,
) {
	if err := c.obsServer.SendEvent(msg.ChannelId, "enableAudio", msg); err != nil {
		return nil, err
	}

	return &emptypb.Empty{}, nil
}
func (c *GrpcImpl) ObsAudioDisable(
	_ context.Context,
	msg *websockets.ObsAudioDisableOrEnableMessage,
) (
	*emptypb.Empty, error,
) {
	if err := c.obsServer.SendEvent(msg.ChannelId, "disableAudio", msg); err != nil {
		return nil, err
	}

	return &emptypb.Empty{}, nil
}
func (c *GrpcImpl) ObsStopStream(
	_ context.Context,
	msg *websockets.ObsStopOrStartStream,
) (*emptypb.Empty, error) {
	if err := c.obsServer.SendEvent(msg.ChannelId, "stopStream", msg); err != nil {
		return nil, err
	}

	return &emptypb.Empty{}, nil
}
func (c *GrpcImpl) ObsStartStream(
	_ context.Context,
	msg *websockets.ObsStopOrStartStream,
) (*emptypb.Empty, error) {
	if err := c.obsServer.SendEvent(msg.ChannelId, "startStream", msg); err != nil {
		return nil, err
	}

	return &emptypb.Empty{}, nil
}

func (c *GrpcImpl) ObsCheckIsUserConnected(
	_ context.Context,
	msg *websockets.ObsCheckUserConnectedRequest,
) (*websockets.ObsCheckUserConnectedResponse, error) {
	res, err := c.obsServer.IsUserConnected(msg.UserId)
	if err != nil {
		return nil, err
	}

	return &websockets.ObsCheckUserConnectedResponse{
		State: res,
	}, nil
}
